wicker/core/datasets.py

```diff
@@ -125,14 +188,166 @@ def __len__(self) -> int:
+        local_path = schema_path
+        # only fetch the file locally if we have a cache prefix
+        if self._local_cache_path_prefix is not None:
+            local_path = self._storage.fetch_file(
+                schema_path, self._local_cache_path_prefix, timeout_seconds=self._filelock_timeout_seconds
+            )
+        with open(local_path, "rb") as f:
+            schema_data = serialization.loads(
+                f.read().decode("utf-8"), treat_objects_as_bytes=self._treat_objects_as_bytes
+            )
+        return schema_data
+
+    @staticmethod
+    def __create_column_bytes_file_reader(
+        dataset_name: Optional[str],
+        filelock_timeout_seconds: int,
+        local_cache_path_prefix: Optional[str],
+        path_factory: WickerPathFactory,
+        storage: AbstractDataStorage,
+    ) -> Union[ColumnBytesFileCache, ColumnBytesFileReader]:
+        """Create a column bytes file reader class.
+
+        Creates an instance of a ColumnBytesFileReader or ColumnBytesFileCache
+        depending on whether a local_cache_path_prefix is not none. Access
+        pattern is the same between the two classes but one is a read through cache
+        and the other is not, usage dependent on if you need your own cache or rely
+        on bucket mount cache in GCSFuse or S3 Mount Point.
+
+        Args:
+            dataset_name (Optional[str]): Name of the dataset.
+            filelock_timeout_seconds (int): Time that you wait for file lock.
+            local_cache_path_prefix (Optional[str]): Optional path of the prefix, determines
+                if you use a cache or not. If specified as None no cache is used.
+            path_factory (WickerPathFactory): Path factory for locating column bytes files
+                in dataset path.
+            storage (AbstractDataStorage): Data storage to use for grabbing files.
+
+        Returns:
+            Union[ColumnBytesFileCache, ColumnBytesFileReader]: reader class for reading data
+            with or without cache.
+        """
+        # if we have a cache prefix create a cache
+        if local_cache_path_prefix is not None:
+            logging.info(f"Cache passed at path - {local_cache_path_prefix}, creating read through cache on top.")
+            # weird mypy problem, can't define baseclass and have it pick up the child correctly
+            column_bytes_file_class: Union[ColumnBytesFileCache, ColumnBytesFileReader] = ColumnBytesFileCache(
+                filelock_timeout_seconds=filelock_timeout_seconds,
+                local_cache_path_prefix=local_cache_path_prefix,
+                path_factory=path_factory,
+                storage=storage,
+            )
+        else:
+            logging.info("No cache passed, reading without caching.")
+            column_bytes_file_class = ColumnBytesFileReader(dataset_name=dataset_name, path_factory=path_factory)
+        return column_bytes_file_class
+
+
+class FileSystemDataset(AbstractDataset):
```

Comment:
Possible yes, for this PR probably not. Like I mentioned above it's out of our scope to do a larger refactor in this PR or maybe ever. I totally agree with what you're saying where a user should just define `Dataset` object but someone borked the original design and that didn't happen. The abstraction I did awhile ago was meant to be a halfway measure. I like what you're suggesting I think it's for a future version of dataset library however.

I'm happy to give you context on why this happened with Wicker but really right now it just needs to work given it's other myriad of problems. It's probably the best path to stop supporting Wicker and redesign it with what we've learned with a similar but improved API. If you're okay with that my team or yours can re-design this if you request or take it on and I'm sure we can collaborate and support it. Both @convexquad and I think this is the correct path and have discussed it.

Reply:
I can also tell from the other PR that was opened what you're getting at but I will mention these are meant to be parallel and not orthogonal. They share the same interfaces directly they are just parallel implementations.

Reply:
Prob a factory method to return a AbstractDataset according to the storage backend is better. Otherwise the The abstractDataset could be a implementation detail instead of the a interface. 
The ideal user experience getDataset(name, version, storage: enum (S3, LocalFile))->AbstractDataset(prob no need change the original AbstractDataset)
the implementation of the getDataset is :
make a BaseDataset  which inherits the AbstractDataset and encapsulate the shared logic as a implementation base instead of interface.
1 get different DataStorage backend, cache etc
2 inject the DataStorage backend, cache etc to the BaseDataset 
 3 return the abstract dataset

Reply:
@zhenyu again I think you are right!!! Good job in the review.

Although I am co-creator of this project, actually my team is not the owner of the project. So we tried super hard to preserve the original S3 classes like `S3Dataset`, `S3DataStorage`, etc. (except that we have to add `WickerPathFactory` base class) and to keep the S3 developer experience the same even if it is not ideal.

If it is ok, we are going to try using local storage from this PR and see if it is useful or not and probably we will learn some new things. Then I think we can refactor the developer experience better in the future.